lib/puppet_x/templates/invoke_ps_command.erb

```diff
@@ -0,0 +1,226 @@
+    }
+
+    public override Guid InstanceId
+    {
+        get { return _hostId; }
+    }
+
+    public override string Name
+    {
+        get { return "PuppetPSHost"; }
+    }
+
+    public override Version Version
+    {
+        get { return new Version(1, 0); }
+    }
+
+    public override PSHostUserInterface UI
+    {
+        get { return null; }
+    }
+
+    public override CultureInfo CurrentCulture
+    {
+        get { return Thread.CurrentThread.CurrentCulture; }
+    }
+
+    public override CultureInfo CurrentUICulture
+    {
+        get { return Thread.CurrentThread.CurrentUICulture; }
+    }
+
+    public override void EnterNestedPrompt()
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void ExitNestedPrompt()
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void NotifyBeginApplication()
+    {
+        return;
+    }
+
+    public override void NotifyEndApplication()
+    {
+        return;
+    }
+
+    public override void SetShouldExit(int exitCode)
+    {
+      this.shouldExit = true;
+      this.exitCode = exitCode;
+    }
+}
+"@
+
+function New-XmlResult
+{
+  param(
+    [Parameter()]$exitcode,
+    [Parameter()]$output,
+    [Parameter()]$errormessage
+  )
+
+  # we make our own xml because ConvertTo-Xml makes hard to parse xml ruby side
+  # and we need to be sure
+  [System.Xml.XmlDocument]$xDoc = New-Object System.Xml.XmlDocument
+  [System.Xml.XmlElement]$xRoot = $xDoc.CreateElement("ReturnResult")
+  $xDoc.AppendChild($xRoot) | Out-Null
+
+  [System.Xml.XmlElement]$xElement = $xDoc.CreateElement("Property")
+  $xElement.SetAttribute('Name','exitcode')
+  $xElement.innerText = "$($exitCode)"
+  $xRoot.AppendChild($xElement) | Out-Null
+
+  if($errormessage -ne $null){
+    $errormessage = [System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($errormessage))
+    [System.Xml.XmlElement]$xElement = $xDoc.CreateElement("Property")
+    $xElement.SetAttribute('Name','errormessage')
+    $xElement.innerText = $errormessage
+    $xRoot.AppendChild($xElement) | Out-Null
+  }
+
+  if($output -ne $null){
+    $output = [System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($output))
+    [System.Xml.XmlElement]$xElement = $xDoc.CreateElement("Property")
+    $xElement.SetAttribute('Name','stdout')
+    $xElement.innerText = $output
+    $xRoot.AppendChild($xElement) | Out-Null
+  }
+
+  $memStream = New-Object System.IO.MemoryStream
+  $writerSettings = New-Object System.Xml.XmlWriterSettings
+  $writerSettings.Indent = $true
+  $writerSettings.Encoding = New-Object System.Text.UTF8Encoding -Arg $false
+  $writerSettings.ConformanceLevel = [System.Xml.ConformanceLevel]::Document
+
+  $writer = [System.Xml.XmlWriter]::Create($memStream, $writerSettings)
+  $xDoc.Save($writer)
+
+  [System.Text.Encoding]::UTF8.GetString($memStream.ToArray())
+}
+
+$cleanupfunction = @"
+function Remove-ScriptVariables($path) {
```

Comment:
In doing research I've seen your suggested approach (store before and compare to after) and my approach (search and find variables). I'm trying to have as few Invoke()'s as possible and avoid throwing exceptions.

Store Before/CompareAfter
pro: Remove-Variable can exclude an array of names, so easy oneliners
cons: At least two seperate Invoke()'s are needed outside of the primary Invoke()

Search and Destroy
pro: uses PowerShell's own parser to find variables and remove (silent if cant). One Invoke()
cons: relies on clever code